contracts-periphery/src/WithdrawHook.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+pragma abicoder v2;
+
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "src/utils/ISwapRouter.sol";
+
+contract WithdrawHook {
+
+  ISwapRouter internal immutable swapRouter;
+  address internal constant WETH9 = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+  constructor(ISwapRouter _swapRouter) {
+  swapRouter = _swapRouter;
+  }
+
+  function tokensWithdrawn(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data
+  ) external{
+    // Plz recommend checks that should happen here
+    _swapForEth(_amount, _stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _data);
+  }
+
+  function _swapForEth(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data) internal {
+    uint256 senderBalance = _amount;
+    if(_sponsorFee > 0) {
+      SafeERC20.safeTransferFrom(IERC20(_tokenAddr), address(this), _sponsor, _sponsorFee);
+      senderBalance -= _sponsorFee;
+    }
+
+    (address _dest, uint256 _swapAmount, uint256 _amountOutMinimum, uint24 _poolFee) = abi.decode(_data, (address, uint256, uint256, uint24));
+    senderBalance -= _swapAmount;
+
+    SafeERC20.safeApprove(IERC20(_tokenAddr), address(swapRouter), _swapAmount);
```

Comment:
Instead of approving the router with every swap, we should consider a method like this that anyone can call to initialize this contract for that token, to save gas for users. We can add an onlyOwner modifier if we're worried about the risk of arbitrary token approvals (though this contract should never hold funds so it should be safe, but need to think on it more)

```solidity
function approveToken(IERC20 _token) external {
  _token.safeApprove(address(swapRouter), type(uint256).max)
}
```

Reply:
Looks like this change was implemented without an `onlyOwner` modifier. I'm pretty sure this should be safe (still haven't spent the time thinking about it more though 😅), but just curious how you decided on this decision?

Reply:
Ended up adding onlyOwner after since the frontend won't include random tokens anyway